README.md
imdbphp2
========

PHP library providing access to IMDB info

doc/CHANGELOG
$Id: CHANGELOG 513 2012-04-15 19:26:57Z izzy $

History for IMDBPHP2
====================

v2.1.3 (15.04.2012)
-------------------
! some redirects have not been handled correctly (i.e. when IMDBID xxxxxxx was re-routed to yyyyyyy)
! imdb::is_serial() and imdb::get_episode_details() were broken due to site changes
! imdb::episodes() was broken due to site changes. ATTENTION: with this rewrite of the method,
  the outer array no longer starts at 0 but reflects the real season number (as often requested)!
+ imdb::get_episode_details() - added "airdate" as requested by ticket:290
! imdb::goofs() and imdb::crazy_credits() where broken due to site changes
* some adjustments to imdb::trivia - spoilers are now avoided by default. Pass TRUE as only argument to retrieve them separately.
+ some enhancements to genre retrieval (thanks to George!) - now there will be more genres reported :)

v2.1.2 (19.01.2012)
-------------------
! site changes broke the imdb class methods keywords(), seasons(), languages(), mpaa_reason(), goofs()
! site changes confused imdb::extReviews(), imdb::videosites()
! company info URLs were just the IMDB relative URL w/o protocol and site. Added the missing info
! if production spanned multiple years (xxxx-yyyy), movie type was setup wrongly ("type xxxx-")
+ added method imdb::endyear() for cases where production spanned multiple years. So for
  "xxxx-yyyy" this returns "yyyy", while imdb::year() is adjusted to "xxxx". Otherwise, both
  return the same year.
! imdb_topcharts::getChartsBoxOffice() was broken. Fixed thanx to hints from razor.
! imdb::episodes() was completely broken, as the entire page design and structure changed (multiple pages now)
+ added imdb::get_episode_details() to retrieve "uplink info" on episodes

v2.1.1 (19.12.2011)
-------------------
+ imdb: adding new budget class
+ imdb nowplaying: added Top-10-Boxoffice
+ movies: added method movietype to check wheter it is a movie, tv series, ...
* updated API reference
! is_serial() did always return FALSE (site-change)
* fixed a bunch of NOTICEs
! imdb::languages() had problems with some ISO settings - hopefully fixed now.
* sometimes imdb::genres() did not return all available genres

v2.1.0 (11.10.2011)
-------------------
! movieposterdb class was broken due to site changes at movieposterdb.com; applied a quick-hack to fix it
! fixes due to site changes: imdb::rating, imdb::goofs
* complete rewrite of imdb_nowplaying class (location and structure of page had changed)
+ some improvements to episode data (see http://projects.izzysoft.de/trac/imdbphp/ticket/255 for details)
! trailers on Moviepilot were broken due to changed URL
! pilot: fixed mainPictures() due to API changes
! pilot_person: fixed photos() due to API changes
+ added basic check for pilot classes (corresponding to the IMDB checks)

v2.0.7 (21.08.2011)
-------------------
! site changes again: fixed imdb_person:name (tnx to GeorgeFive and Qvist)
! same reason: imdb::votes(), imdb::country() and imdb::mpaa_reason()
! same reason, just a couple of days later: imdb rating
! imdb_person class: filmografies no longer listed movie names due to site changes
* imdb_person class: character names w/o id are listed again in filmografies
! IMDB started to sometimes pre- or postfix movie titles with "IMDb", becoming part of the title then
* imdb movie: plotoutline() did not work for unrated movies

v2.0.6 (08.07.2011)
-------------------
! next IMDB site changes: ratings
! imdb_person filmographies: quarters in years, e.g. 1924/II, lead to empty year
* imdb::awards(): minor adjustments by Qvist
! imdb_movielist: fixed the listing methods (were all broken by IMDB site changes some time ago)
+ movies (imdb+pilot): Added new method languages_detailed() to get detailed lang info (name, code and comment of languages used)
+ imdb::episodes(): Result array now also includes season and episode (numbers)
+ movie classes: Added new method aspect_ratio to obtain the screen aspect ratio

v2.0.5 (25.05.2011)
-------------------
! IMDB-Link for person search in demo was wrong
+ movie class: new method yearspan() for e.g. serials spanning multiple production years
! IMDB site changes broke actors filmographies, quotes, user comments, trivia, top250
* adaption to imdb::storyline() as suggested by an anonymous user

v2.0.4 (28.12.2010)
-------------------
! mdb_request.class.php ignored the IMDBPHP_CONFIG constant - fixed
! they did it again - so imdb::comment_split() was broken
! imdb photo stuff was broken
+ runtimes and genres know have an additional source as fallback
! imdb::votes() was broken due to site changes
! imdb::seasons() always returned 1 (if there were seasons) for IMDB switched order
* cleaned up imdb::prodNotes() result
* cleaning up double-paragraphs in imdb::plotoutline()
! person.php linked back to imdb.php instead of movie.php

v2.0.3 (17.10.2010)
-------------------
! some after-fixes for the last "big run" include imdb methods plotoutline(),
  runtime_all()
! same for imdb_person::name()
* imdb::plotoutline() was taken from storyline with the recent changes - fixed it up
+ but since we had a working storyline, that method was added as storyline()
! imdb::mainPictures() was also broken by the last site changes - fixed
! mdb_request class did not inherit force_agent and trigger_referer values from
  mdb_config and thus issued PHP notices

v2.0.2 (10.10.2010)
-------------------
! imdbsearch: some search results missed the year (if it was consisting of more than a plain
  year, e.g. something like "2001/II")
! imdb::prodNotes: lastUpdate was broken due to a site change
+ movies: added imdb::locations() to retrieve details on the filming locations.
  pilot::locations() is just a wrapper to this requiring full fallback configured.
! mpaa_reason returned wrong text (tnx to dblume for report and fix!)
+ added method is_serial() to the movie classes to determine whether the selected
  entry is part of a TV series

test/index.php
<?php
 #############################################################################
 # IMDBPHP                              (c) Giorgos Giagas & Itzchak Rehberg #
 # written by Giorgos Giagas                                                 #
 # extended & maintained by Itzchak Rehberg <izzysoft AT qumran DOT org>     #
 # http://www.izzysoft.de/                                                   #
 # ------------------------------------------------------------------------- #
 # This program is free software; you can redistribute and/or modify it      #
 # under the terms of the GNU General Public License (see doc/LICENSE)       #
 #############################################################################

 /* $Id: index.php 479 2011-10-12 08:58:51Z izzy $ */

#===============================================[ optional URL Parameters ]===
# check: only check the specified class (movie, name, charts, nowplay, trail)
#        default: Check them all
# skip : skip the test for the specified, comma-separated classes. Has
#        priority over "check"
# cron : enable cron-mode (cron=1) - i.e. output plain text, and on errors only
# cache: enable (1|on|true) or disable (0|off|false) caching. Overrides config.

require_once(dirname(__FILE__)."/helpers.inc");
#=====================================[ Make sure it works out-of-the-box ]===
$os = php_uname('s');
$path = ini_get('include_path');
if (strtolower(substr($os,0,3))=="win") ini_set('include_path',".;..;$path");
else ini_set('include_path',".:..:$path");

#====================================================[ Output HTML Header ]===
if (!CRON) {
  echo "<!DOCTYPE HTML PUBLIC '-//W3C//DTD HTML 4.01 Transitional//EN'>\n";
  echo "<HTML><HEAD>\n"
     . " <TITLE>IMDBAPI Checks</TITLE>\n"
     . " <STYLE TYPE='text/css'>body,td,th { font-size:12px; font-family:sans-serif; }</STYLE>\n"
     . "</HEAD><BODY>\n";
}
#==========================================[ Define what should be tested ]===
$check_movie   = false;
$check_name    = false;
$check_charts  = false;
$check_nowplay = false;
$check_budget  = false;
$uri = $_SERVER['REQUEST_URI'];
if ( strpos($uri,'?')>0 ) $uri .= '&amp;';
else $uri .= '?';
raw("<SPAN STYLE='margin-right:2em;'><A HREF='{$uri}api=imdb'>IMDB</A></SPAN><SPAN STYLE='margin-left:2em;'><A HREF='{$uri}api=pilot'>PILOT</A></SPAN>\n");
if ( empty($_REQUEST['api']) && !in_array($_REQUEST['api'],array('imdb','pilot')) ) {
  if (CRON) $api = 'imdb';
  else {
    raw("</BODY></HTML>\n");
    exit;
  }
} else {
  $api = $_REQUEST['api'];
}
switch ($_REQUEST["check"]) {
  case "movie"   : $check_movie   = true; break;
  case "name"    : $check_name    = true; break;
  case "charts"  : $check_charts  = true; break;
  case "nowplay" : $check_nowplay = true; break;
  case "budget"  : $check_budget  = true; break;
  default:
    $check_movie   = true;
    $check_name    = true;
    $check_charts  = true;
    $check_nowplay = true;
    $check_budget  = true;
}
if (!empty($_REQUEST["skip"])) {
  $skips = explode(",",$_REQUEST["skip"]);
  $segments = array("movie","name","charts","nowplay");
  foreach($skips as $skip) ${"check_$skip"} = FALSE;
}

#=========================================================[ Run the tests ]===
if ($check_movie)   { if ($api=='imdb') require('imdb.inc'); else require('pilot.inc'); }
if ($check_name)    { if ($api=='imdb') require('imdb_person.inc'); else require ('pilot_person.inc'); }
if ($check_charts)  { if ($api=='imdb') require ("imdb_charts.inc"); }
if ($check_nowplay) { if ($api=='imdb') require ("imdb_nowplaying.inc"); }
if ($check_budget)  { if ($api=='imdb') require ("imdb_budget.inc"); }

#===============================================[ Summary and HTML footer ]===
$passed  = $methods - $failures;
$percent = round(100*$passed/$methods)."%";
if (CRON) {
  if ($failures>0 || !empty($failed)) {
    echo "Some methods failed the automated tests:\n"
       . "========================================\n\n";
    foreach ($failed as $fail) echo "* $fail\n";
    echo "\nOverall results:\n\n"
       . "Success: $passed/$methods ($percent)\n";
  }
} else {
  headline("Test Results:");
  raw("<UL><LI>Methods: $methods</LI><LI>Passed: $passed</LI><LI>Failures: $failures</LI><LI>Success: $percent</LI></UL>\n");
  if ($percent == "100%") raw("<b>Congratulations!</b> Looks like the complete API is working perfectly.");
  raw("</BODY></HTML>\n");
}
?>

movie_base.class.php
<?php
 #############################################################################
 # IMDBPHP.MoviePilot                                    (c) Itzchak Rehberg #
 # written by Itzchak Rehberg <izzysoft AT qumran DOT org>                   #
 # http://www.izzysoft.de/                                                   #
 # ------------------------------------------------------------------------- #
 # This program is free software; you can redistribute and/or modify it      #
 # under the terms of the GNU General Public License (see doc/LICENSE)       #
 #############################################################################

 /* $Id: movie_base.class.php 497 2012-01-18 07:05:19Z izzy $ */

 require_once (dirname(__FILE__)."/browseremulator.class.php");
 require_once (dirname(__FILE__)."/mdb_base.class.php");
 require_once (dirname(__FILE__)."/mdb_request.class.php");

#===================================================[ The Movie Base class ]===
/** Accessing Movie information
 * @package MDBApi
 * @class movie_base
 * @extends mdb_base
 * @author Izzy (izzysoft AT qumran DOT org)
 * @copyright (c) 2009 by Itzchak Rehberg and IzzySoft
 * @version $Revision: 497 $ $Date: 2012-01-18 08:05:19 +0100 (Mi, 18. Jan 2012) $
 */
class movie_base extends mdb_base {

#------------------------------------------------------------[ Constructor ]---
  /** Initialize class
   * @constructor movie_base
   * @param string id IMDBID to use for data retrieval
   */
  function __construct ($id) {
    parent::__construct($id);
  }

#--------------------------------------------------[ Start (over) / Reset ]---
  /** Reset page vars
   * @method protected reset_vars
   */
  protected function reset_vars() {
   $this->page["Title"] = "";
   $this->page["TitleFoot"] = ""; // IMDB only, as part of info was outsourced
   $this->page["Credits"] = "";
   $this->page["CrazyCredits"] = "";
   $this->page["Amazon"] = "";
   $this->page["Goofs"] = "";
   $this->page["Trivia"] = "";
   $this->page["Plot"] = "";
   $this->page["Synopsis"] = "";
   $this->page["Comments"] = "";
   $this->page["Quotes"] = "";
   $this->page["Taglines"] = "";
   $this->page["Plotoutline"] = "";
   $this->page["Trivia"] = "";
   $this->page["Directed"] = "";
   $this->page["Episodes"] = "";
   $this->page["Quotes"] = "";
   $this->page["Trailers"] = "";
   $this->page["MovieConnections"] = "";
   $this->page["ExtReviews"] = "";
   $this->page["ReleaseInfo"] = "";
   $this->page["CompanyCredits"] = "";
   $this->page["ParentalGuide"] = "";
   $this->page["OfficialSites"] = "";
   $this->page["Keywords"] = "";
   $this->page["Awards"] = "";
   $this->page["Locations"] = "";

   $this->akas = array();
   $this->awards = array();
   $this->countries = array();
   $this->castlist = array(); // pilot only
   $this->crazy_credits = array();
   $this->credits_cast = array();
   $this->credits_composer = array();
   $this->credits_director = array();
   $this->credits_producer = array();
   $this->credits_writing = array();
   $this->extreviews = array();
   $this->goofs = array();
   $this->langs = array();
   $this->langs_full = array();
   $this->aspectratio = "";
   $this->main_comment = "";
   $this->main_genre = "";
   $this->main_keywords = array();
   $this->all_keywords = array();
   $this->main_language = "";
   $this->main_photo = "";
   $this->main_thumb = "";
   $this->main_pictures = array();
   $this->main_plotoutline = "";
   $this->main_rating = -1;
   $this->main_runtime = "";
   $this->main_movietype = "";
   $this->main_title = "";
   $this->main_votes = -1;
   $this->main_year = -1;
   $this->main_endyear = -1;
   $this->main_yearspan = array();
   $this->main_creator = array();
   $this->main_tagline = "";
   $this->main_storyline = "";
   $this->main_prodnotes = array();
   $this->main_movietypes = array();
   $this->main_top250 = -1;
   $this->moviecolors = array();
   $this->movieconnections = array();
   $this->moviegenres = array();
   $this->moviequotes = array();
   $this->movieruntimes = array();
   $this->mpaas = array();
   $this->mpaas_hist = array();
   $this->mpaa_justification = "";
   $this->plot_plot = array();
   $this->synopsis_wiki = "";
   $this->release_info = array();
   $this->seasoncount = -1;
   $this->season_episodes = array();
   $this->sound = array();
   $this->soundtracks = array();
   $this->split_comment = array();
   $this->split_plot = array();
   $this->taglines = array();
   $this->trailers = array();
   $this->video_sites = array();
   $this->trivia = array();
   $this->compcred_prod = array();
   $this->compcred_dist = array();
   $this->compcred_special = array();
   $this->compcred_other = array();
   $this->parental_guide = array();
   $this->official_sites = array();
   $this->locations = array();
  }


} // end class movie_base
?>


movie.php
<?php
 #############################################################################
 # IMDBPHP                              (c) Giorgos Giagas & Itzchak Rehberg #
 # written by Giorgos Giagas                                                 #
 # extended & maintained by Itzchak Rehberg <izzysoft AT qumran DOT org>     #
 # http://www.izzysoft.de/                                                   #
 # ------------------------------------------------------------------------- #
 # This program is free software; you can redistribute and/or modify it      #
 # under the terms of the GNU General Public License (see doc/LICENSE)       #
 #############################################################################

 /* $Id: movie.php 510 2012-04-15 18:08:14Z izzy $ */

if (isset ($_GET["mid"])) {
  $movieid = $_GET["mid"];
  $engine  = $_GET["engine"];

  switch($engine) {
    case "pilot":
        require("pilot.class.php");
        $movie = new pilot($_GET["mid"]);
        $charset = "utf8";
        $source  = "<A HREF='?engine=imdb&mid=$movieid'>IMDB</A> | <B CLASS='active'>MoviePilot</B>";
        if ($movie->get_pilot_imdbfill()) $source .= '<SUP>+i</SUP>';
        break;
    default:
        require("imdb.class.php");
        $movie = new imdb($_GET["mid"]);
        $charset = "iso-8859-1";
        $source  = "<B CLASS='active'>IMDB</B> | <A HREF='?engine=pilot&mid=$movieid'>MoviePilot</A>";
        break;
  }

  $movie->setid ($movieid);
  $rows = 2; // count for the rowspan; init with photo + year

  echo "<!DOCTYPE HTML PUBLIC '-//W3C//DTD HTML 4.01 Transitional//EN'>\n";
  echo "<HTML><HEAD>\n";
  echo " <TITLE>".$movie->title().' ('.$movie->year().") [IMDBPHP2 v".$movie->version." Demo]</TITLE>\n";
  echo " <STYLE TYPE='text/css'>body,td,th { font-size:12px; font-family:sans-serif; } b.active { color:#b00;background-color:#fff;text-decoration:underline;}</STYLE>\n";
  echo " <META http-equiv='Content-Type' content='text/html; charset=$charset'>\n";
  echo "</HEAD>\n<BODY ONLOAD='fix_colspan()'>\n<TABLE BORDER='1' ALIGN='center' STYLE='border-collapse:collapse'>";

  # Title & year
  echo '<TR><TH COLSPAN="3" STYLE="background-color:#ffb000">';
  echo "[IMDBPHP2 v".$movie->version." Demo] Movie Details for '" . $movie->title()."' (".$movie->year().")";
  echo "<SPAN STYLE='float:right;text-align:right;display:inline !important;font-size:75%;'>Source: [$source]</SPAN>";
  echo "</TH></TR>\n";
  flush();

  # Photo
  echo '<TR><TD id="photocol" rowspan="29" valign="top">';
  if (($photo_url = $movie->photo_localurl() ) != FALSE) {
    echo '<img src="'.$photo_url.'" alt="Cover">';
  } else {
    echo "No photo available";
  }

  # AKAs
  $aka = $movie->alsoknow();
  $cc  = count($aka);
  if (!empty($aka)) {
    echo '</TD><TD valign=top width=120><b>Also known as:</b> </td><td>';
    foreach ( $aka as $ak){
      echo $ak["title"];
      if (!empty($ak["year"])) echo " ".$ak["year"];
      echo  " =&gt; ".$ak["country"];
      if (empty($ak["lang"])) { if (!empty($ak["comment"])) echo " (".$ak["comment"].")"; }
      else {
        if (!empty($ak["comment"])) echo ", ".$ak["comment"];
        echo " [".$ak["lang"]."]";
      }
      echo "<BR>";
    }
    echo "</td></tr>\n";
    flush();
  }

  # Movie Type
  ++$rows;
  echo '<TR><TD><B>Type:</B></TD><TD>'.$movie->movietype()."</TD></TR>\n";

  # Keywords
  $keywords = $movie->keywords();
  if ( !empty($keywords) ) {
    ++$rows;
    echo '<TR><TD><B>Keywords:</B></TD><TD>'.implode(', ',$keywords)."</TD></TR>\n";
  }

  # Seasons
  if ( $movie->seasons() != 0 ) {
    ++$rows;
    echo '<TR><TD><B>Seasons:</B></TD><TD>'.$movie->seasons()."</TD></TR>\n";
    flush();
  }

  # Episode Details
  $ser = $movie->get_episode_details();
  if (!empty($ser)) {
    ++$rows;
    echo '<TR><TD><B>Series Details:</B></TD><TD>'.$ser['seriestitle'].' Season '.$ser['season'].', Episode '.$ser['episode'].", Airdate ".$ser['airdate']."</TD></TR>\n";
  }

  # Year & runtime
  echo '<TR><TD><B>Year:</B></TD><TD>'.$movie->year().'</TD></TR>';
  $runtime = $movie->runtime();
  if (!empty($runtime)) {
    ++$rows;
    echo "<TR><TD valign=top><B>Runtime:</b></TD><TD>$runtime minutes</TD></TR>\n";
  }
  flush();

  # MPAA
  $mpaa = $movie->mpaa();
  if (!empty($mpaa)) {
    ++$rows;
    $mpar = $movie->mpaa_reason();
    if (empty($mpar)) echo '<TR><TD><B>MPAA:</b></TD><TD>';
    else echo '<TR><TD rowspan="2"><B>MPAA:</b></TD><TD>';
    echo "<table align='left' border='1' style='border-collapse:collapse;background-color:#ddd;'><tr><th style='background-color:#07f;'>Country</th><th style='background-color:#07f;'>Rating</th></tr>";
    foreach ($mpaa as $key=>$mpaa) {
      echo "<tr><td>$key</td><td>$mpaa</td></tr>";
    }
    echo "</table></TD></TR>\n";
    if (!empty($mpar)) {
      ++$rows;
      echo "<TR><TD>$mpar</TD></TR>\n";
    }
  }

  # Ratings and votes
  $ratv = $movie->rating();
  if (!empty($ratv)) { echo "<TR><TD><B>Rating:</b></TD><TD>$ratv</TD></TR>\n"; ++$rows; }
  $ratv = $movie->votes();
  if (!empty($ratv)) { echo "<TR><TD><B>Votes:</B></TD><TD>$ratv</TD></TR>\n"; ++$rows; }
  flush();

  # Languages
  $languages = $movie->languages();
  if (!empty($languages)) {
    ++$rows;
    echo '<TR><TD><B>Languages:</B></TD><TD>';
    for ($i = 0; $i + 1 < count($languages); $i++) {
      echo $languages[$i].', ';
    }
    echo $languages[$i]."</TD></TR>\n";
  }
  flush();

  # Country
  $country = $movie->country();
  if (!empty($country)) {
    ++$rows;
    echo '<TR><TD><B>Country:</B></TD><TD>';
    for ($i = 0; $i + 1 < count($country); $i++) {
      echo $country[$i].', ';
    }
    echo $country[$i]."</TD></TR>\n";
  }

  # Genres
  $genre = $movie->genre();
  if (!empty($genre)) { echo "<TR><TD><B>Genre:</B></TD><TD>$genre</TD></TR>\n"; ++$rows; }

  $gen = $movie->genres();
  if (!empty($gen)) {
    ++$rows;
    echo '<TR><TD><B>All Genres:</B></TD><TD>';
    for ($i = 0; $i + 1 < count($gen); $i++) {
      echo $gen[$i].', ';
    }
    echo $gen[$i]."</TD></TR>\n";
  }

  # Colors
  $col = $movie->colors();
  if (!empty($col)) {
    ++$rows;
    echo '<TR><TD><B>Colors:</B></TD><TD>';
    for ($i = 0; $i + 1 < count($col); $i++) {
      echo $col[$i].', ';
    }
    echo $col[$i]."</TD></TR>\n";
  }
  flush();

  # Sound
  $sound = $movie->sound ();
  if (!empty($sound)) {
    ++$rows;
    echo '<TR><TD><B>Sound:</B></TD><TD>';
    for ($i = 0; $i + 1 < count($sound); $i++) {
      echo $sound[$i].', ';
    }
    echo $sound[$i]."</TD></TR>\n";
  }

  $tagline = $movie->tagline();
  if (!empty($tagline)) {
    ++$rows;
    echo "<TR><TD valign='top'><B>Tagline:</B></TD><TD>$tagline</TD></TR>\n";
  }

  #==[ Staff ]==
  # director(s)
  $director = $movie->director();
  if (!empty($director)) {
    ++$rows;
    echo '<TR><TD valign=top><B>Director:</B></TD><TD>';
    echo "<table align='left' border='1' style='border-collapse:collapse;background-color:#ddd;'><tr><th style='background-color:#07f;'>Name</th><th style='background-color:#07f;'>Role</th></tr>";
    for ($i = 0; $i < count($director); $i++) {
      echo '<tr><td width=200>';
      echo "<a href='person.php?engine=$engine&mid=".$director[$i]["imdb"]."'>";
      echo $director[$i]["name"].'</a></td><td>';
      echo $director[$i]["role"]."</td></tr>";
    }
    echo "</table></td></tr>\n";
  }

  # Story
  $write = $movie->writing();
  if (!empty($write)) {
    ++$rows;
    echo '<TR><TD valign=top><B>Writing By:</B></TD><TD>';
    echo "<table align='left' border='1' style='border-collapse:collapse;background-color:#ddd;'><tr><th style='background-color:#07f;'>Name</th><th style='background-color:#07f;'>Role</th></tr>";
    for ($i = 0; $i < count($write); $i++) {
      echo '<tr><td width=200>';
      echo "<a href='person.php?engine=$engine&mid=".$write[$i]["imdb"]."'>";
      echo $write[$i]["name"].'</a></td><td>';
      echo $write[$i]["role"]."</td></tr>";
    }
    echo "</table></td></tr>\n";
  }
  flush();

  # Producer
  $produce = $movie->producer();
  if (!empty($produce)) {
    ++$rows;
    echo '<TR><TD valign=top><B>Produced By:</B></TD><TD>';
    echo "<table align='left' border='1' style='border-collapse:collapse;background-color:#ddd;'><tr><th style='background-color:#07f;'>Name</th><th style='background-color:#07f;'>Role</th></tr>";
    for ($i = 0; $i < count($produce); $i++) {
      echo '<tr><td width=200>';
      echo "<a href='person.php?engine=$engine&mid=".$produce[$i]["imdb"]."'>";
      echo $produce[$i]["name"].'</a></td><td>';
      echo $produce[$i]["role"]."</td></tr>";
    }
    echo "</table></td></tr>\n";
  }

  # Music
  $compose = $movie->composer();
  if (!empty($compose)) {
    ++$rows;
    echo '<TR><TD valign=top><B>Music:</B></TD><TD>';
    echo "<table align='left' border='1' style='border-collapse:collapse;background-color:#ddd;'><tr><th style='background-color:#07f;'>Name</th><th style='background-color:#07f;'>Role</th></tr>";
    for ($i = 0; $i < count($compose); $i++) {
      echo '<tr><td width=200>';
      echo "<a href='person.php?engine=$engine&mid=".$compose[$i]["imdb"]."'>";
      echo $compose[$i]["name"]."</a></td></tr>";
    }
    echo "</table></td></tr>\n";
  }
  flush();

  # Cast
  $cast = $movie->cast();
  if (!empty($cast)) {
    ++$rows;
    echo '<TR><TD valign=top><B>Cast:</B></TD><TD>';
    echo "<table align='left' border='1' style='border-collapse:collapse;background-color:#ddd;'><tr><th style='background-color:#07f;'>Actor</th><th style='background-color:#07f;'>Role</th></tr>";
    for ($i = 0; $i < count($cast); $i++) {
      echo '<tr><td width=200>';
      echo "<a href='person.php?engine=$engine&mid=".$cast[$i]["imdb"]."'>";
      echo $cast[$i]["name"].'</a></td><td>';
      echo $cast[$i]["role"]."</td></tr>";
    }
    echo "</table></td></tr>\n";
  }
  flush();

  # Plot outline & plot
  $plotoutline = $movie->plotoutline();
  if (!empty($plotoutline)) {
    ++$rows;
    echo "<tr><td valign='top'><b>Plot Outline:</b></td><td>$plotoutline</td></tr>\n";
  }

  $plot = $movie->plot();
  if (!empty($plot)) {
    ++$rows;
    echo '<tr><td valign=top><b>Plot:</b></td><td><ul>';
    for ($i = 0; $i < count($plot); $i++) {
      echo "<li>".$plot[$i]."</li>\n";
    }
    echo "</ul></td></tr>\n";
  }
  flush();

  # Taglines
  $taglines = $movie->taglines();
  if (!empty($taglines)) {
    ++$rows;
    echo '<tr><td valign=top><b>Taglines:</b></td><td><ul>';
    for ($i = 0; $i < count($taglines); $i++) {
      echo "<li>".$taglines[$i]."</li>\n";
    }
    echo "</ul></td></tr>\n";
  }

  # Seasons
  if ( $movie->is_serial() || $movie->seasons() ) {
    ++$rows;
    $episodes = $movie->episodes();
    echo '<tr><td valign=top><b>Episodes:</b></td><td>';
    foreach ( $episodes as $season => $ep ) {
      foreach ( $ep as $episodedata ) {
        echo '<b>Season '.$episodedata['season'].', Episode '.$episodedata['episode'].': <a href="'.$_SERVER["PHP_SELF"].'?mid='.$episodedata['imdbid'].'">'.$episodedata['title'].'</a></b> (<b>Original Air Date: '.$episodedata['airdate'].'</b>)<br>'.$episodedata['plot'].'<br/><br/>';
      }
    }
    echo "</td></tr>\n";
  }

  # Locations
  $locs = $movie->locations();
  if (!empty($locs)) {
    ++$rows;
    echo '<tr><td valign="top"><b>Filming Locations:</b></td><td><ul>';
    foreach ($locs as $loc) {
      if ( empty($loc['url']) ) echo '<li>'.$loc['name'].'</li>';
      else echo '<li><a href="http://'.$movie->imdbsite.$loc['url'].'">'.$loc['name'].'</a></li>';
    }
    echo "</ul></td></tr>\n";
  }

  # Selected User Comment
  $comment = $movie->comment();
  if (!empty($comment)) {
    ++$rows;
    echo "<tr><td valign='top'><b>User Comments:</b></td><td>$comment</td></tr>\n";
  }

  # Quotes
  $quotes = $movie->quotes();
  if (!empty($quotes)) {
    ++$rows;
    echo '<tr><td valign=top><b>Movie Quotes:</b></td><td>';
    echo preg_replace("/http\:\/\/".str_replace(".","\.",$movie->imdbsite)."\/name\/nm(\d{7})\//","person.php?engine=$engine&mid=\\1",$quotes[0])."</td></tr>\n";
  }

  # Trailer
  $trailers = $movie->trailers(TRUE);
  if (!empty($trailers)) {
    ++$rows;
    echo '<tr><td valign=top><b>Trailers:</b></td><td>';
    for ($i=0;$i<count($trailers);++$i) {
      if (!empty($trailers[$i]['url'])) echo "<a href='".$trailers[$i]['url']."'>".$trailers[$i]['title']."</a><br>\n";
    }
    echo "</td></tr>\n";
  }

  # Crazy Credits
  $crazy = $movie->crazy_credits();
  $cc    = count($crazy);
  if ($cc) {
    ++$rows;
    echo '<tr><td valign=top><b>Crazy Credits:</b></td><td>';
    echo "We know about $cc <i>Crazy Credits</i>. One of them reads:<br>$crazy[0]</td></tr>\n";
  }

  # Goofs
  $goofs = $movie->goofs();
  $gc    = count($goofs);
  if ($gc > 0) {
    ++$rows;
    echo '<tr><td valign=top><b>Goofs:</b></td><td>';
    echo "We know about $gc goofs. Here comes one of them:<br>";
    echo "<b>".$goofs[0]["type"]."</b> ".$goofs[0]["content"]."</td></tr>\n";
  }

  # Trivia
  $trivia = $movie->trivia();
  $gc     = count($trivia);
  if ($gc > 0) {
    ++$rows;
    echo '<tr><td valign=top><b>Trivia:</b></td><td>';
    echo "There are $gc entries in the trivia list - like these:<br><ul>";
    for ($i=0;$i<5;++$i) {
      if (empty($trivia[$i])) break;
      echo "<li>".preg_replace("/http\:\/\/".str_replace(".","\.",$movie->imdbsite)."\/name\/nm(\d{7})\//","person.php?engine=$engine&mid=\\1",$trivia[$i])."</li>";
    }
    echo "</ul></td></tr>\n";
  }

  # Soundtracks
  $soundtracks = $movie->soundtrack();
  $gc = count($soundtracks);
  if ($gc > 0) {
    ++$rows;
    echo '<tr><td valign=top><b>Soundtracks:</b></td><td>';
    echo "There are $gc soundtracks listed - like these:<br>";
    echo "<table align='center' border='1' style='border-collapse:collapse;background-color:#ddd;'><tr><th style='background-color:#07f;'>Soundtrack</th><th style='background-color:#07f;'>Credit 1</th><th style='background-color:#07f;'>Credit 2</th></tr>";
    for ($i=0;$i<5;++$i) {
      if (empty($soundtracks[$i])) break;
      $credit1 = preg_replace("/http\:\/\/".str_replace(".","\.",$movie->imdbsite)."\/name\/nm(\d{7})\//","person.php?engine=$engine&mid=\\1",$soundtracks[$i]["credits"][0]);
      $credit2 = preg_replace("/http\:\/\/".str_replace(".","\.",$movie->imdbsite)."\/name\/nm(\d{7})\//","person.php?engine=$engine&mid=\\1",$soundtracks[$i]["credits"][1]);
      echo "<tr><td>".$soundtracks[$i]["soundtrack"]."</td><td>$credit1</td><td>$credit2</td></tr>";
    }
    echo "</table></td></tr>\n";
  }

  echo "</TABLE><BR>\n";
  echo "<SCRIPT TYPE='text/javascript'>// <!--\n";
  echo "  function fix_colspan() {\n";
  echo "    document.getElementById('photocol').rowSpan = '$rows';\n";
  echo "  }\n//-->\n</SCRIPT>\n";
  echo "</BODY></HTML>";
}
?>
